File: sdks/sdkwork-craw-chat-sdk/sdkwork-craw-chat-sdk-flutter/composed/lib/src/conversations_module.dart
import 'package:backend_sdk/backend_sdk.dart';

import 'builders.dart';
import 'context.dart';
import 'types.dart';

class CrawChatConversationsModule {
  final CrawChatSdkContext context;

  CrawChatConversationsModule(this.context);

  Future<CreateConversationResult?> create(CreateConversationRequest body) {
    return context.backendClient.conversation.createConversation(body);
  }

  Future<CreateConversationResult?> createAgentDialog(
    CreateAgentDialogRequest body,
  ) {
    return context.backendClient.conversation.createAgentDialog(body);
  }

  Future<CreateConversationResult?> createAgentHandoff(
    CreateAgentHandoffRequest body,
  ) {
    return context.backendClient.conversation.createAgentHandoff(body);
  }

  Future<CreateConversationResult?> createSystemChannel(
    CreateSystemChannelRequest body,
  ) {
    return context.backendClient.conversation.createSystemChannel(body);
  }

  Future<ConversationSummaryView?> get(String conversationId) {
    return context.backendClient.conversation.getConversationSummary(conversationId);
  }

  Future<AgentHandoffStateView?> getAgentHandoffState(String conversationId) {
    return context.backendClient.conversation.getAgentHandoffState(conversationId);
  }

  Future<AgentHandoffStateView?> acceptAgentHandoff(String conversationId) {
    return context.backendClient.conversation.acceptAgentHandoff(conversationId);
  }

  Future<AgentHandoffStateView?> resolveAgentHandoff(String conversationId) {
    return context.backendClient.conversation.resolveAgentHandoff(conversationId);
  }

  Future<AgentHandoffStateView?> closeAgentHandoff(String conversationId) {
    return context.backendClient.conversation.closeAgentHandoff(conversationId);
  }

  Future<ListMembersResponse?> listMembers(String conversationId) {
    return context.backendClient.conversation.listConversationMembers(conversationId);
  }

  Future<ConversationMember?> addMember(
    String conversationId,
    AddConversationMemberRequest body,
  ) {
    return context.backendClient.conversation.addConversationMember(
      conversationId,
      body,
    );
  }

  Future<ConversationMember?> removeMember(
    String conversationId,
    RemoveConversationMemberRequest body,
  ) {
    return context.backendClient.conversation.removeConversationMember(
      conversationId,
      body,
    );
  }

  Future<TransferConversationOwnerResult?> transferOwner(
    String conversationId,
    TransferConversationOwnerRequest body,
  ) {
    return context.backendClient.conversation.transferConversationOwner(
      conversationId,
      body,
    );
  }

  Future<ChangeConversationMemberRoleResult?> changeMemberRole(
    String conversationId,
    ChangeConversationMemberRoleRequest body,
  ) {
    return context.backendClient.conversation.changeConversationMemberRole(
      conversationId,
      body,
    );
  }

  Future<ConversationMember?> leave(String conversationId) {
    return context.backendClient.conversation.leave(conversationId);
  }

  Future<ConversationReadCursorView?> getReadCursor(String conversationId) {
    return context.backendClient.conversation.getConversationReadCursor(
      conversationId,
    );
  }

  Future<ConversationReadCursorView?> updateReadCursor(
    String conversationId,
    UpdateReadCursorRequest body,
  ) {
    return context.backendClient.conversation.updateConversationReadCursor(
      conversationId,
      body,
    );
  }

  Future<TimelineResponse?> listMessages(String conversationId) {
    return context.backendClient.conversation.listConversationMessages(
      conversationId,
    );
  }

  Future<PostMessageResult?> postMessage(
    String conversationId,
    PostMessageRequest body,
  ) {
    return context.backendClient.conversation.postConversationMessage(
      conversationId,
      body,
    );
  }

  Future<PostMessageResult?> postText(
    String conversationId, {
    required String text,
    CrawChatTextMessageOptions options = const CrawChatTextMessageOptions(),
  }) {
    return postMessage(
      conversationId,
      CrawChatBuilders.textMessage(text: text, options: options),
    );
  }

  Future<PostMessageResult?> publishSystemMessage(
    String conversationId,
    PostMessageRequest body,
  ) {
    return context.backendClient.conversation.publishSystemChannelMessage(
      conversationId,
      body,
    );
  }

  Future<PostMessageResult?> publishSystemText(
    String conversationId, {
    required String text,
    CrawChatTextMessageOptions options = const CrawChatTextMessageOptions(),
  }) {
    return publishSystemMessage(
      conversationId,
      CrawChatBuilders.textMessage(text: text, options: options),
    );
  }
}

File: sdks/sdkwork-craw-chat-sdk/sdkwork-craw-chat-sdk-typescript/composed/src/conversations-module.ts
import { buildTextMessageRequest } from './builders.js';
import type {
  AddConversationMemberRequest,
  AgentHandoffStateView,
  ChangeConversationMemberRoleRequest,
  ChangeConversationMemberRoleResult,
  ConversationMember,
  ConversationReadCursorView,
  ConversationSummaryView,
  CreateAgentDialogRequest,
  CreateAgentHandoffRequest,
  CreateConversationRequest,
  CreateConversationResult,
  CreateSystemChannelRequest,
  ListMembersResponse,
  PostMessageRequest,
  PostMessageResult,
  PostTextMessageOptions,
  RemoveConversationMemberRequest,
  TimelineResponse,
  TransferConversationOwnerRequest,
  TransferConversationOwnerResult,
  UpdateReadCursorRequest,
} from './types.js';
import type { CrawChatSdkContext } from './sdk-context.js';

export class CrawChatConversationsModule {
  constructor(private readonly context: CrawChatSdkContext) {}

  create(body: CreateConversationRequest): Promise<CreateConversationResult> {
    return this.context.backendClient.conversation.createConversation(body);
  }

  createAgentDialog(body: CreateAgentDialogRequest): Promise<CreateConversationResult> {
    return this.context.backendClient.conversation.createAgentDialog(body);
  }

  createAgentHandoff(
    body: CreateAgentHandoffRequest,
  ): Promise<CreateConversationResult> {
    return this.context.backendClient.conversation.createAgentHandoff(body);
  }

  createSystemChannel(
    body: CreateSystemChannelRequest,
  ): Promise<CreateConversationResult> {
    return this.context.backendClient.conversation.createSystemChannel(body);
  }

  get(conversationId: string | number): Promise<ConversationSummaryView> {
    return this.context.backendClient.conversation.getConversationSummary(
      conversationId,
    );
  }

  getAgentHandoffState(
    conversationId: string | number,
  ): Promise<AgentHandoffStateView> {
    return this.context.backendClient.conversation.getAgentHandoffState(
      conversationId,
    );
  }

  acceptAgentHandoff(
    conversationId: string | number,
  ): Promise<AgentHandoffStateView> {
    return this.context.backendClient.conversation.acceptAgentHandoff(
      conversationId,
    );
  }

  resolveAgentHandoff(
    conversationId: string | number,
  ): Promise<AgentHandoffStateView> {
    return this.context.backendClient.conversation.resolveAgentHandoff(
      conversationId,
    );
  }

  closeAgentHandoff(
    conversationId: string | number,
  ): Promise<AgentHandoffStateView> {
    return this.context.backendClient.conversation.closeAgentHandoff(
      conversationId,
    );
  }

  listMembers(conversationId: string | number): Promise<ListMembersResponse> {
    return this.context.backendClient.conversation.listConversationMembers(
      conversationId,
    );
  }

  addMember(
    conversationId: string | number,
    body: AddConversationMemberRequest,
  ): Promise<ConversationMember> {
    return this.context.backendClient.conversation.addConversationMember(
      conversationId,
      body,
    );
  }

  removeMember(
    conversationId: string | number,
    body: RemoveConversationMemberRequest,
  ): Promise<ConversationMember> {
    return this.context.backendClient.conversation.removeConversationMember(
      conversationId,
      body,
    );
  }

  transferOwner(
    conversationId: string | number,
    body: TransferConversationOwnerRequest,
  ): Promise<TransferConversationOwnerResult> {
    return this.context.backendClient.conversation.transferConversationOwner(
      conversationId,
      body,
    );
  }

  changeMemberRole(
    conversationId: string | number,
    body: ChangeConversationMemberRoleRequest,
  ): Promise<ChangeConversationMemberRoleResult> {
    return this.context.backendClient.conversation.changeConversationMemberRole(
      conversationId,
      body,
    );
  }

  leave(conversationId: string | number): Promise<ConversationMember> {
    return this.context.backendClient.conversation.leave(conversationId);
  }

  getReadCursor(
    conversationId: string | number,
  ): Promise<ConversationReadCursorView> {
    return this.context.backendClient.conversation.getConversationReadCursor(
      conversationId,
    );
  }

  updateReadCursor(
    conversationId: string | number,
    body: UpdateReadCursorRequest,
  ): Promise<ConversationReadCursorView> {
    return this.context.backendClient.conversation.updateConversationReadCursor(
      conversationId,
      body,
    );
  }

  listMessages(conversationId: string | number): Promise<TimelineResponse> {
    return this.context.backendClient.conversation.listConversationMessages(
      conversationId,
    );
  }

  postMessage(
    conversationId: string | number,
    body: PostMessageRequest,
  ): Promise<PostMessageResult> {
    return this.context.backendClient.conversation.postConversationMessage(
      conversationId,
      body,
    );
  }

  postText(
    conversationId: string | number,
    text: string,
    options: PostTextMessageOptions = {},
  ): Promise<PostMessageResult> {
    return this.postMessage(conversationId, buildTextMessageRequest(text, options));
  }

  publishSystemMessage(
    conversationId: string | number,
    body: PostMessageRequest,
  ): Promise<PostMessageResult> {
    return this.context.backendClient.conversation.publishSystemChannelMessage(
      conversationId,
      body,
    );
  }

  publishSystemText(
    conversationId: string | number,
    text: string,
    options: PostTextMessageOptions = {},
  ): Promise<PostMessageResult> {
    return this.publishSystemMessage(
      conversationId,
      buildTextMessageRequest(text, options),
    );
  }
}
